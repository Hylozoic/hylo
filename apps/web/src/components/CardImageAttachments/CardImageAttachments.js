import cx from 'classnames'
import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { filter, isEmpty } from 'lodash/fp'
import ModalDialog from 'components/ModalDialog'
import ImageCarousel from 'components/ImageCarousel'
import classes from './CardImageAttachments.module.scss'

export default function CardImageAttachments ({
  attachments,
  className,
  isFlagged
}) {
  const imageAttachments = filter({ type: 'image' }, attachments)

  const firstImageUrl = imageAttachments?.[0]?.url
  const otherImageUrls = imageAttachments?.slice(1).map(ia => ia.url)

  const [initialSlide, setInitialSlide] = useState(0)
  const [modalVisible, setModalVisible] = useState(false)

  const toggleModal = (e) => {
    if (className === 'post-card') return
    setInitialSlide(e?.target.dataset.index || 0)
    setModalVisible(!modalVisible)
  }

  const modalSettings = {
    showCancelButton: false,
    submitButtonText: 'Close',
    showModalTitle: false,
    closeModal: toggleModal,
    style: { width: '100%', maxWidth: '1024px' }
  }

  if (isEmpty(imageAttachments)) return null
  if (!firstImageUrl) return null

  return (
    <>
      <div className={cx(className, classes.image, { [classes.flagged]: isFlagged })}>
        <img
          src={firstImageUrl}
          alt='Attached image 1'
          data-index={0}
          onClick={toggleModal}
          data-testid='first-image'
        />
        <div className={classes.others}>
          <div className={classes.othersInner}>
            {!isEmpty(otherImageUrls) && otherImageUrls.map((url, index) =>
              <img
                className={classes.other}
                data-index={index + 1}
                src={url}
                alt={`Attached image ${index + 2}`}
                key={index}
                onClick={toggleModal}
              />
            )}
          </div>
        </div>
      </div>
      {modalVisible && (
        <ModalDialog {...modalSettings}>
          <ImageCarousel attachments={imageAttachments} initialSlide={initialSlide} />
        </ModalDialog>
      )}
    </>
  )
}

CardImageAttachments.propTypes = {
  attachments: PropTypes.array.isRequired,
  linked: PropTypes.bool,
  className: PropTypes.string
}
