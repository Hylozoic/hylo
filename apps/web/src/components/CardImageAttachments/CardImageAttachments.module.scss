.images {
  position: relative;

  img {
    cursor: pointer;
  }
}

.others {
  position: absolute;
  width: 100%;
  bottom: 15px;
  right: 0;
  display: flex;
  overflow-x: auto;
  padding: 10px 0;
  padding-left: $space-2x;
  cursor: pointer;
}

.others-inner {
  display: flex;
  flex-direction: row;
  margin-left: auto;
}

.other {
  display: block;
  border: 2px solid #fff;
  width: auto;
  height: 80px;
  border-radius: 3px;
  margin-right: 20px;
  background-size: cover;
  box-shadow: 0px 4px 10px rgba(42, 64, 89, 0.6);

  &:hover {
    filter: brightness(110%);
  }
}

.chatPost.images {
  display: flex;
  overflow-x: auto;
  overflow-y: hidden;
  height: 90px;
  margin-left: 20px;
}

.chatPost {
  .imagesInner {
    display: flex;
    flex-direction: row;
    //margin-left: auto;
  }

  .image {
    display: block;
    width: 80px;
    height: 80px;
    cursor: pointer;
    border-radius: 3px;
    margin-right: 20px;
    background-size: cover;
    border: 1px solid $color-cape-cod-10;
    box-shadow: 0px 4px 10px rgba(42, 64, 89, 0.3);

    &:hover {
      filter: brightness(110%);
    }
  }
}

.flagged {
  overflow: hidden;

  img {
    filter: blur(30px);
  }
}
