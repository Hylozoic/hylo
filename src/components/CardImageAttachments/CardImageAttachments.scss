.image {
  background-position: center;
  background-size: cover;
  @include fluid-height(62%);
  width: 100%;
  margin: 0 auto $space-2x auto;
}

.link {
  width: 100%;
  height: 100%;
  display: block;
  position: absolute;
}

.others {
  position: absolute;
  width: 100%;
  bottom: 0.6rem;
  display: flex;
  overflow-x: auto;
  padding-left: $space-2x;
}

.others-inner {
  display: flex;
  flex-direction: row;
  margin-left: auto;
}

.other {
  display: block;
  > div {
    width: 80px;
    height: 80px;
    border-radius: 3px;
    margin-right: 0.6rem;
    background-size: cover;
    border: 1px solid $color-cape-cod-10;
    &:hover {
      filter: brightness(110%);
    }
  }
}
