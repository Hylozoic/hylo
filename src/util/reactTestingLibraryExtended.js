import React from 'react'
import { MemoryRouter } from 'react-router'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import rootReducer from 'store/reducers'
import createMiddleware from 'store/middleware'

export function generateStore (history, initialState) {
  return createStore(rootReducer, initialState, createMiddleware(history))
}

export const AllTheProviders = (store) => ({ children }) => {
  return (
    <Provider store={store || createStore(() => {})}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  )
}

// re-export everything

const rtl = require('react-testing-library')
const customRender = (ui, options, providersFunc) =>
  rtl.render(ui, { wrapper: providersFunc, ...options })

module.exports = {
  ...rtl,
  render: customRender
}
