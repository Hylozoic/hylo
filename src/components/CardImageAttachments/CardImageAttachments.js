import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { filter, isEmpty } from 'lodash/fp'
import ModalDialog from 'components/ModalDialog'
import ImageCarousel from 'components/ImageCarousel'
import './CardImageAttachments.scss'

export default function CardImageAttachments ({
  attachments,
  className
}) {
  const imageAttachments = filter({ type: 'image' }, attachments)

  if (isEmpty(imageAttachments)) return null

  const firstImageUrl = imageAttachments[0].url
  const otherImageUrls = imageAttachments.slice(1).map(ia => ia.url)

  if (!firstImageUrl) return null

  const [modalVisible, setModalVisible] = useState(false)
  const toggleModal = () => {
    setModalVisible(!modalVisible)
  }

  const modalSettings = {
    showCancelButton: false,
    submitButtonText: 'Close',
    showModalTitle: false,
    closeModal: toggleModal,
    style: { width: '100%', maxWidth: '1024px' }
  }

  return (
    <>
      <div className={className} styleName='image'>
        <img src={firstImageUrl} onClick={toggleModal} />
        <div styleName='others'>
          <div styleName='others-inner'>
            {!isEmpty(otherImageUrls) && otherImageUrls.map(url =>
              <img styleName='other' src={url} onClick={toggleModal} />
            )}
          </div>
        </div>
      </div>
      {modalVisible && <ModalDialog {...modalSettings}>
        <ImageCarousel attachments={imageAttachments} />
      </ModalDialog>}
    </>
  )
}

CardImageAttachments.propTypes = {
  attachments: PropTypes.array.isRequired,
  linked: PropTypes.bool,
  className: PropTypes.string
}
