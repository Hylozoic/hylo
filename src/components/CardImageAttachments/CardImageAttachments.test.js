import React from 'react'
import CardImageAttachments from './CardImageAttachments'
import { shallow } from 'enzyme'
import { render, screen } from 'util/testing/reactTestingLibraryExtended'
import userEvent from '@testing-library/user-event'

it('renders no images', () => {
  expect(shallow(<CardImageAttachments attachments={[
    { url: 'bonkerz', type: 'file' },
    { url: 'bonkers', type: 'file' },
    { url: 'bonkerzztop', type: 'file' }
  ]} />)).toEqual({})
})

it('renders a single image', () => {
  expect(shallow(<CardImageAttachments attachments={[
    { url: 'foo', type: 'image' },
    { url: 'bonkerz', type: 'file' }
  ]} />)).toMatchSnapshot()
})

it('renders multiple images', () => {
  expect(shallow(<CardImageAttachments attachments={[
    { url: 'bar', type: 'image' },
    { url: 'baz', type: 'image' },
    { url: 'bonk', type: 'image' },
    { url: 'bonkerz', type: 'file' }
  ]} />)).toMatchSnapshot()
})

it('displays modal when image is clicked', async () => {
  render(<CardImageAttachments attachments={[
    { url: 'bar', type: 'image' },
    { url: 'baz', type: 'image' },
    { url: 'bonk', type: 'image' },
    { url: 'bonkerz', type: 'file' }
  ]} />)

  expect(await screen.findByAltText('Attached image 1')).toBeInTheDocument()
  expect(await screen.findByAltText('Attached image 2')).toBeInTheDocument()
  expect(await screen.findByAltText('Attached image 3')).toBeInTheDocument()
  expect(await screen.queryByAltText('Attached image 4')).not.toBeInTheDocument()

  userEvent.click(await screen.findByAltText('Attached image 2'))
  expect(await screen.findByTestId('sc-img0')).toBeInTheDocument()
  expect(await screen.findByTestId('sc-img1')).toBeInTheDocument()
  expect(await screen.findByTestId('sc-img2')).toBeInTheDocument()

  // implementation dependent, but the only way to test that the active slide is image 2
  expect((await screen.findByTestId('sc-img0')).parentNode.parentNode).toHaveAttribute('aria-hidden', 'true')
  expect((await screen.findByTestId('sc-img1')).parentNode.parentNode).toHaveAttribute('aria-hidden', 'false')
  expect((await screen.findByTestId('sc-img2')).parentNode.parentNode).toHaveAttribute('aria-hidden', 'true')
})
