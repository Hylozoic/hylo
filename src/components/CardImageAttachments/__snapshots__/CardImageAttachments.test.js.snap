// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a single image 1`] = `
<Fragment>
  <div
    data-stylename="image"
  >
    <img
      onClick={[Function]}
      src="foo"
    />
    <div
      data-stylename="others"
    >
      <div
        data-stylename="others-inner"
      />
    </div>
  </div>
</Fragment>
`;

exports[`renders multiple images 1`] = `
<Fragment>
  <div
    data-stylename="image"
  >
    <img
      onClick={[Function]}
      src="bar"
    />
    <div
      data-stylename="others"
    >
      <div
        data-stylename="others-inner"
      >
        <img
          data-stylename="other"
          onClick={[Function]}
          src="baz"
        />
        <img
          data-stylename="other"
          onClick={[Function]}
          src="bonk"
        />
      </div>
    </div>
  </div>
</Fragment>
`;
